File: LeetCode Problems/801 to 900/872. Leaf-Similar Trees/872. Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<int> list;
    void helper(TreeNode *root) {
        if (root == nullptr)
            return;
        if (root->left == nullptr && root->right == nullptr) {
            list.push_back(root->val);
        }
        helper(root->left);
        helper(root->right);
    }

public:
    bool leafSimilar(TreeNode *root1, TreeNode *root2) {
        helper(root1);
        vector<int> list1 = list;
        list.clear();
        helper(root2);
        if (list.size() != list1.size())
            return false;
        for (int i = 0; i < list.size(); i++)
            if (list[i] != list1[i])
                return false;
        return true;
    }
};

File: LeetCode Problems/801 to 900/872. Leaf-Similar Trees/872. Leaf-Similar Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> list = new ArrayList<Integer>();
    private void helper(TreeNode root) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            list.add(root.val);
        }
        helper(root.left);
        helper(root.right);
    }
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        helper(root1);
        List<Integer> list1 = list;
        list = new ArrayList<Integer>();
        helper(root2);
        if (list1.size() != list.size()) return false;
        for (int i = 0; i < list.size(); i++) {
            if (list1.get(i) != list.get(i)) return false;
        }
        return true;
    }
}
